### PR TITLE
Add glTF camera support in glTFExporter

### LINKED ISSUE #9146: Add camera serialization to GLTF serializer
See:
https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#cameras

### PR DESCRIPTION
**Pull request by daoshengmu**
Fixes #9146.

Per glTF camera spec and tutorial [1], I added `perspective` and `orthographic` cameras and `orientation` support in the exporter.

[1] https://github.com/KhronosGroup/glTF-Tutorials/blob/master/gltfTutorial/gltfTutorial_016_Cameras.md

### CODE REVIEW

**Comment by daoshengmu** [comment]
After these changes, camera in glTF file would look like below

```
"nodes": [
{
      "name": "Camera1",
      "translation": [
        0,
        100,
        100
      ],
      "rotation": [
        -0.17024385059860536,
        0.8788524074865445,
        -0.08475879752424419,
        0.43755162373222783
      ],
      "camera": 0
    }
],
"cameras": [
    {
      "type": "perspective",
      "name": "Camera1",
      "perspective": {
        "aspectRatio": 0.9789719626168224,
        "yfov": 0.8,
        "znear": 1,
        "zfar": 10000
      }
    }
 ],
```

File: serializers/src/glTF/2.0/glTFExporter.ts [hunk 1]
```diff
@@ -1182,6 +1191,22 @@ export class _Exporter {
...
     }
 
+    private setCameraTransformation(node: INode, babylonCamera: Camera, convertToRightHandedSystem: boolean): void {
+        if (!babylonCamera.position.equalsToFloats(0, 0, 0)) {
+            node.translation = convertToRightHandedSystem ? _GLTFUtilities._GetRightHandedPositionVector3(babylonCamera.position).asArray() : babylonCamera.position.asArray();
+        }
+
+        let rotation = babylonCamera.absoluteRotation;
+        let rotationQuaternion = Quaternion.RotationYawPitchRoll(rotation.y, rotation.x, rotation.z);
+        if (!(rotationQuaternion.x === 0 && rotationQuaternion.y === 0 && rotationQuaternion.z === 0 && rotationQuaternion.w === 1)) {
```
Comment (bghgary) [inline serializers/src/glTF/2.0/glTFExporter.ts:1201]: nit:
```suggestion
        if (!Quaternion.IsIdentity(rotationQuaternion)) {
```

File: serializers/src/glTF/2.0/glTFExporter.ts [hunk 2]
```diff
@@ -1696,6 +1721,34 @@ export class _Exporter {
...
+            };
+
+            if (camera.name) {
+                glTFCamera.name = camera.name;
+            }
+
+            if (glTFCamera.type === CameraType.PERSPECTIVE) {
+                glTFCamera.perspective = {aspectRatio: camera.getEngine().getAspectRatio(camera),
+                    yfov: camera._cache.fovMode === Camera.FOVMODE_VERTICAL_FIXED ? camera.fov : camera.fov * camera._cache.aspectRatio,
+                    znear: camera.minZ, zfar: camera.maxZ};
```
Comment (bghgary) [inline serializers/src/glTF/2.0/glTFExporter.ts:1737]: nit:
```suggestion
                glTFCamera.perspective = {
                    aspectRatio: camera.getEngine().getAspectRatio(camera),
                    yfov: camera._cache.fovMode === Camera.FOVMODE_VERTICAL_FIXED ? camera.fov : camera.fov * camera._cache.aspectRatio,
                    znear: camera.minZ, zfar: camera.maxZ
                };
```

File: serializers/src/glTF/2.0/glTFExporter.ts [hunk 3]
```diff
@@ -1721,6 +1774,16 @@ export class _Exporter {
                                 }
                             }
 
+                            if (babylonNode instanceof Camera) {
+                                // Assign camera index by comparing with node names.
+                                for (let i = 0; i < this._cameras.length; i++) {
+                                    let babylonCamera: Camera = babylonNode;
+                                    if (this._cameras[i].name == babylonCamera.name) {
```
Comment (bghgary) [inline serializers/src/glTF/2.0/glTFExporter.ts:1781]: I'm not sure this will work all the time. `babylonCamera.name` is not guaranteed to be unique.

**Review by bghgary — COMMENTED** [review]
Looks good to me. Thanks for the contribution. Just a nit and a comment.

File: serializers/src/glTF/2.0/glTFExporter.ts [hunk 4]
```diff
@@ -1182,6 +1191,22 @@ export class _Exporter {
         }
     }
 
+    private setCameraTransformation(node: INode, babylonCamera: Camera, convertToRightHandedSystem: boolean): void {
+        if (!babylonCamera.position.equalsToFloats(0, 0, 0)) {
+            node.translation = convertToRightHandedSystem ? _GLTFUtilities._GetRightHandedPositionVector3(babylonCamera.position).asArray() : babylonCamera.position.asArray();
+        }
+
+        let rotation = babylonCamera.absoluteRotation;
+        let rotationQuaternion = Quaternion.RotationYawPitchRoll(rotation.y, rotation.x, rotation.z);
```
Comment (bghgary) [inline serializers/src/glTF/2.0/glTFExporter.ts:1200]: Is this right? `absoluteRotation` is already a Quaternion.

File: serializers/src/glTF/2.0/glTFExporter.ts (hunk 3 above)
Comment (daoshengmu) [reply serializers/src/glTF/2.0/glTFExporter.ts]: I use a hashMap to store its index instead. Let me know how it looks to you.

File: serializers/src/glTF/2.0/glTFExporter.ts (hunk 4 above)
Comment (daoshengmu) [reply serializers/src/glTF/2.0/glTFExporter.ts]: Good catch!

**Review by deltakosh — APPROVED** [review]

**Review by bghgary — APPROVED** [review]

**Pull request merged**